### PR TITLE
Add ProviderManager, make provider deletion generic

### PR DESCRIPTION
**Pull request by dmjb**
Relates to #2845

Change the `ProviderFactory` to `ProviderManager`. Add new method for deleting provider instances. This hides provider-specific cleanup behind an appropriate interface, and removes some duplication of logic (with slightly different implementations) in different parts of the codebase.

Most of the code changes relate to wiring this new logic into the rest of the codebase and replacing what was there.

# Summary

***Provide a brief overview of the changes and the issue being addressed. 
Explain the rationale and any background necessary for understanding the changes. 
List dependencies required by this change, if any.***

Fixes #(related issue)

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [x] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

File: internal/projects/delete.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ func CleanUpUnmanagedProjects(
 	proj uuid.UUID,
 	querier db.Querier,
 	authzClient authz.Client,
-	providerService service.GitHubProviderService,
+	providerManager manager.ProviderManager,
```
Comment (dmjb) [inline internal/projects/delete.go:41]: In future, I might investigate creating a `ProjectService` or similar - right now we end up having to pass around a bunch of dependencies through the controlplane for the project creation logic.

**Comment by eleftherias** [comment]
This is nice! I haven't tested manually, but the code is looking good. Is there anything in particular you're looking for to take this out of draft?

**Comment by dmjb** [comment]
@eleftherias I need to add some test coverage, and also do some manual verification of the changes.

File: internal/controlplane/handlers_providers_test.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ import (
 	"github.com/google/uuid"
```
Comment (dmjb) [inline internal/controlplane/handlers_providers_test.go:22]: I decided to reuse these test cases to test the deletion codepath in `GitHubProviderClassManager`

**Review by eleftherias — APPROVED** [review]
Ran some manual testing, everything is working as expected.

**Pull request merged**